libsql-sqlite3/src/vectorIndex.c

```diff
@@ -765,82 +844,94 @@ int vectorIndexCreate(Parse *pParse, Index *pIdx, const char *zDbSName, const Id
   if( pIdx->pPartIdxWhere != NULL ) {
     sqlite3ErrorMsg(pParse, "partial vector index is not supported");
-    return SQLITE_ERROR;
+    goto fail;
   }
 
   pArgsList = pIdx->aColExpr->a[0].pExpr->x.pList;
   pListItem = pArgsList->a;
 
   if( pArgsList->nExpr < 1 ){
     sqlite3ErrorMsg(pParse, VECTOR_INDEX_MARKER_FUNCTION " must contain at least one argument");
-    return SQLITE_ERROR;
+    goto fail;
   }
   if( pListItem[0].pExpr->op != TK_COLUMN ) {
     sqlite3ErrorMsg(pParse, VECTOR_INDEX_MARKER_FUNCTION " first argument must be a column token");
-    return SQLITE_ERROR;
+    goto fail;
   }
   iEmbeddingColumn = pListItem[0].pExpr->iColumn;
   if( iEmbeddingColumn < 0 ) {
     sqlite3ErrorMsg(pParse, VECTOR_INDEX_MARKER_FUNCTION " first argument must be column with vector type");
-    return SQLITE_ERROR;
+    goto fail;
   }
   assert( iEmbeddingColumn >= 0 && iEmbeddingColumn < pTable->nCol );
 
   zEmbeddingColumnTypeName = sqlite3ColumnType(&pTable->aCol[iEmbeddingColumn], "");
   if( vectorIdxParseColumnType(zEmbeddingColumnTypeName, &type, &dims, &pzErrMsg) != 0 ){
     sqlite3ErrorMsg(pParse, "%s: %s", pzErrMsg, zEmbeddingColumnTypeName);
-    return SQLITE_ERROR;
+    goto fail;
   }
 
   // schema is locked while db is initializing and we need to just proceed here
   if( db->init.busy == 1 ){
-    goto succeed;
+    goto ok;
   }
 
   rc = initVectorIndexMetaTable(db, zDbSName);
   if( rc != SQLITE_OK ){
-    return rc;
+    sqlite3ErrorMsg(pParse, "failed to init vector index meta table: %s", sqlite3_errmsg(db));
+    goto fail;
   }
   rc = parseVectorIdxParams(pParse, &idxParams, type, dims, pListItem + 1, pArgsList->nExpr - 1);
   if( rc != SQLITE_OK ){
-    return rc;
+    sqlite3ErrorMsg(pParse, "failed to parse vector idx params");
+    goto fail;
   }
   if( vectorIdxKeyGet(pTable, &idxKey, &pzErrMsg) != 0 ){
     sqlite3ErrorMsg(pParse, "failed to detect underlying table key: %s", pzErrMsg);
-    return SQLITE_ERROR;
+    goto fail;
   }
   if( idxKey.nKeyColumns != 1 ){
     sqlite3ErrorMsg(pParse, "vector index for tables without ROWID and composite primary key are not supported");
-    return SQLITE_ERROR;
+    goto fail;
   }
   rc = diskAnnCreateIndex(db, zDbSName, pIdx->zName, &idxKey, &idxParams);
   if( rc != SQLITE_OK ){
     sqlite3ErrorMsg(pParse, "unable to initialize diskann vector index");
-    return rc;
+    goto fail;
   }
   rc = insertIndexParameters(db, zDbSName, pIdx->zName, &idxParams);
+  if( rc == SQLITE_CONSTRAINT ){
+    // we are violating unique constraint here which means that someone inserted parameters in the table before us
+    // taking aside corruption scenarios, this can be in case of loading dump (because tables are loaded before indices) or vacuum-ing DB
+    // both these cases are valid and we must proceed with index creating but avoid index-refill step as it is already filled
+    goto skip_refill;
```

Comment:
Yes, here they are unnecessary - let me remove them.
But usually, this is how I do resource management in the C (and I code in C since end of the June of this year...).

So, when I need non-trivial resource management I usually use only `goto` in the function body in order to not forget to free something (which is very easy when you have a return in the middle and want to add some code above this statement)

Reply:
I'm used to gotos for error handling, and I'm fine with that, it's just that unstructured control flow can be nasty, so seeing a lot of labels makes me wonder if we are not abusing them :)